core/src/main/scala/org/apache/spark/sql/internal/SnappySessionState.scala

```diff
@@ -628,19 +675,11 @@ class SnappySessionState(val snappySession: SnappySession)
               throw new AnalysisException("Cannot update partitioning/key column " +
                   s"of the table for $colName (among [${nonUpdatableColumns.mkString(", ")}])")
             }
-            // cast the update expressions if required
-            val newExpr = if (attr.dataType.sameType(expr.dataType)) {
-              expr
-            } else {
-              // avoid unnecessary copy+cast when inserting DECIMAL types
-              // into column table
-              expr.dataType match {
-                case _: DecimalType
-                  if attr.dataType.isInstanceOf[DecimalType] => expr
-                case _ => Alias(Cast(expr, attr.dataType), attr.name)()
-              }
+            if (!attr.dataType.sameType(expr.dataType)) {
+              throw new AnalysisException("Implicit type casting is not performed for update" +
+                  " statements")
```

Comment:
As we discussed, this will lead to frequent DECIMAL type mismatches throwing exceptions so user will have to add precision/scala casts unnecessarily. If you add some tests with decimal types then it should immediately show the problem (in set clause and where clause). Implicit cast for DECIMAL types should be retained IMO.

Reply:
Done changes in this logic as follows: 

trying to assign value of some different type which doesn't match the column's data type will fail except for following cases:  

- assigning null value
- assigning narrower decimal to a wider decimal
- assigning narrower numeric type to wider numeric type as far as precision is not compromised